juju/juju
# [JUJU-2749] upgrade a resource without a new charm revision available.

Allow for upgrading a resource even if the charm used by an application will not be upgraded. Three small changes were needed to accomplish this. 

The refresh (update-charm) cli logic on failures related to ErrAlreadyUpToDate had to change. It's not an immediate failure. If there are resources to update or a charm's channel to change, still call SetCharm. Let the user know what is happening.

Update the logic around determination of whether a resource requires an upgrade by using the resource data available in the repository. When upgrading a resource required a charm revision change, this wasn't as important. However with the first change, ensure that we only pass along resources that must be updated. The logic in this part is complex.

SetCharm must to allow for pending resource resolution independent of a charm url change. Including incrementing CharmModifiedVersion if it wasn't done earlier in the method. CharmModifiedVersion is what triggers the uniter to call the upgrade charm hook if the charm url didn't change.

Finally added more integration tests to cover:

1. Refresh a resource without a new charm revision.
2. Refresh a resource, specifying a different revision than deployed, without a new charm revision.
3. Deploy with an uploaded resource, refresh fails to refresh the resource as it's been uploaded.

## QA steps

Depending on which version of ubuntu your machine is running, you may need to run `juju model-default default-series=focal`. Some of these tests use Xenial by default. 

Local charm scenarios:

1. Deploy a local charm without resources. Edit the metadata.yaml file to add a resource. Refresh the charm without uploading a resource file. It should fail.
2. Refresh the local charm from scenario 1 again, this time supplying the resource on the cli: --resource <resource-name>=<path-to-file>. It should succeed.

Use `juju resources <application-name> --format json | jq` to validate resources

```sh
(cd tests ; ./main.sh -v resources)
```

Test with k8s charms and oct-images as well, my attempt was here: https://bugs.launchpad.net/juju/+bug/1993319/comments/3

## Bug reference

Along the road to fixing: https://bugs.launchpad.net/juju/+bug/1993319 . Resources couldn't be updated without a charm revision change before this at all. The next step will be to ensure the bug related to upgrading oci-images is fixed.
